lib/recursive-methods.rb

```diff
@@ -1,49 +1,120 @@
 # Authoring recursive algorithms. Add comments including time and space complexity for each method.
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n)
+
+# factorial(n) Write a method factorial that accepts an integer
+# parameter n and that uses recursion to compute and return the
+# value of n factorial (also known as n!).
+
+#     e.g. fact(4) = 4 * 3 * 2 * 1 = 24
+
 def factorial(n)
-    raise NotImplementedError, "Method not implemented"
+  if (n < 0)
+    raise ArgumentError
+  elsif n == 1 || n == 0
+    return 1
+  else
+    return n * factorial(n - 1)
+  end
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n)
+
+# reverse(s) Write a method reverse that accepts a string as a
+# parameter and then returns the reverse of the string by
+# reversing all letters and all words in the string.
+
+# e.g. reverse("hello world") will return "dlrow olleh"
 def reverse(s)
```

Comment:
👍 
This works, but because you create a new array with each recursive call this is O(n<sup>2</sup>) for both time/space complexity.